controllers/machine_controller.go

```diff
@@ -601,6 +621,60 @@ func (r *MachineReconciler) shouldAdopt(m *clusterv1.Machine) bool {
 	return metav1.GetControllerOf(m) == nil && !util.HasOwner(m.OwnerReferences, clusterv1.GroupVersion.String(), []string{"Cluster"})
 }
 
+func (r *MachineReconciler) watchClusterNodes(ctx context.Context, cluster *clusterv1.Cluster) error {
+	// If there is no tracker, don't watch remote nodes
+	if r.Tracker == nil {
+		return nil
+	}
+
+	if err := r.Tracker.Watch(ctx, remote.WatchInput{
+		Name:         "machine-watchNodes",
+		Cluster:      util.ObjectKey(cluster),
+		Watcher:      r.controller,
+		Kind:         &corev1.Node{},
+		EventHandler: handler.EnqueueRequestsFromMapFunc(r.nodeToMachine),
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *MachineReconciler) nodeToMachine(o client.Object) []reconcile.Request {
+	node, ok := o.(*corev1.Node)
+	if !ok {
+		panic(fmt.Sprintf("Expected a Node but got a %T", o))
+	}
+
+	machineList := &clusterv1.MachineList{}
+	if err := r.Client.List(
+		context.TODO(),
+		machineList,
+		client.MatchingFields{clusterv1.MachineNodeNameIndex: node.Name},
```

Comment:
I didn't catch this in the 0.3 PR, but I just noticed that we're not restricting to machines in a specific namespace or for a specific cluster. We should probably do that, to avoid any possible false positives (e.g. 2 clusters in 2 different VPCs that have the same node name for a node).

```go
func (r *MachineReconciler) nodeToMachine(namespace, clusterName string) func(o client.Object) []reconcile.Request {
  return func(o client.Object) []reconcile.Request {
    // use namespace and clusterName in r.Client.List
  }
}
```

WDYT? cc @vincepri @fabriziopandini

Reply:
This makes sense, it helps with potential conflicts, although it needs to be tested if we can mix and match field matchers with normal ones

Reply:
You mean InNamespace + MatchingFields? Any reason to think that wouldn't work?

Reply:
Looking at the CacheReader implementation, it looks like:

1. If there is a field selector, respect that
1. Else, if there is a namespace match, use that
1. Else, list everything

THEN, if there is a label selector, honor that.

So we can't do namespace + field selector 😦

Reply:
Maybe we should enhance this in C-R?

Reply:
See https://github.com/kubernetes-sigs/controller-runtime/blob/5c2b42d0dfe264fe1a187dcb11f384c0d193c042/pkg/cache/internal/cache_reader.go#L97-L112

and then

https://github.com/kubernetes-sigs/controller-runtime/blob/5c2b42d0dfe264fe1a187dcb11f384c0d193c042/pkg/cache/internal/cache_reader.go#L133-L134

Reply:
Yes, this seems like it could be a good improvement there

Reply:
Do we want to file an issue and merge this PR as-is for now given that's just forward porting?

Reply:
Yes